File: java/Practice01/src/myClient/MySender.java
package myClient;

public class MySender {
	private final MyReceiver _receiver = new MyReceiver();
	
	public MySender(String uri) { }
	
	public MyReceiver getReceiver() {
		return this._receiver;
	}
	
	public void addQuery(int queryId) {
		if (Math.random() < 0.05) {
            throw new MyClientException("Error occurred when add query.");
        }
		
		this._receiver.addQuery(queryId);
	}
	
	public void removeQuery(int queryId) {
		if (Math.random() < 0.05) {
            throw new MyClientException("Error occurred when remove query.");
        }
		
		this._receiver.removeQuery(queryId);
	}
	
	public void close() {
		this._receiver.close();
	}
}

File: csharp/Practice01/src/MyClient/MySender.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace MyClient
{
    public sealed class MySender 
    {
        private readonly Random _random = new Random(DateTime.Now.Millisecond);

        private readonly MyReceiver _receiver = new MyReceiver();

        public MySender(string uri) { }

        public MyReceiver Receiver { get { return this._receiver; } }

        public void AddQuery(int queryId)
        {
            if (this._random.NextDouble() < 0.05)
            {
                throw new MyClientException("Error occurred when add query.");
            }

            this._receiver.AddQuery(queryId);
        }

        public void RemoveQuery(int queryId)
        {
            if (this._random.NextDouble() < 0.05)
            {
                throw new MyClientException("Error occurred when remove query.");
            }

            this._receiver.RemoveQuery(queryId);
        }

        public void Close()
        {
            this._receiver.Close();
        }
    }
}
